File: Two_Pointers/C++/42_trapping_rain_water.cpp
/*
* RUNTIME: 7 ms - Beats 93.57% of submissions
* MEMORY - 22.20 MB - Beats 6.36% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
#include <set>
using namespace std;


/* Thoughtprocess:
* 1. initailize a left and right pointer
* 2. int max left and max right
* 3. initiate while loop while left < right
* 4. evaluate the new max values
* 5. if max_left is less than max_right, add the difference between max_left and height[left] to maxArea
* 6. else add the difference between max_right and height[right] to maxArea
* 7. Depending on the higher value decrease/increase the left/right pointer
* 8. Return the maximum area
* 
*/







class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0, right = height.size() - 1;
        int max_left = 0, max_right = 0;
        int maxArea = 0;

        while (left < right)
        {
            max_left = max(max_left, height[left]);
            max_right = max(max_right, height[right]);
            if (max_left < max_right)
            {
                maxArea += max_left - height[left];
                left++;
            }
            else
            {
                maxArea += max_right - height[right];
                right--;
            }
        } 
        return maxArea;
    }
};

File: Two_Pointers/Python/42_trapping_rain_water.py
# RUNTIME: 101 ms, faster than 80.77% of python3 Submissions. 
# MEMORY USAGE: 18.52 MB, less than 65.02% of python3 Submissions.

from __future__ import annotations
from ast import *


# My thougtprocess:
# 1. Especially after doing the C++ Version I recognize that this could be done more efficiently
# 2. Nonetheless, I want to present my first Bare bones solution that fitted the problem

# 1. initiale left, right, max values and a check
# 2. Checks Purpsose is to not include left and right borders, as the water could not be trapped there
# 3. if left height is smaller than right height, we need to increase left
# 4. return the max area as in the previous two pointer solutions

class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) == 1:
            return 0
        left, right = 0, len(height) - 1
        max_left, max_right = 0, 0
        maxArea = 0
        c = 0
        c2 = 0

        while left < right:
            if height[left] <= c2:
                c2 = height[left]
                left += 1
            else:
                c = 1
        c2 = height[right]
        
        while c == 1 and left < right:
            if height[right] >= c2:
                c2 = height[right]
                right -= 1
            else:
                c = 0
                right += 1
        
        while left < right:
            max_left = max(max_left, height[left])
            max_right = max(max_right, height[right])
            if max_left < max_right:
                maxArea += max_left - height[left]
                left += 1
            else:
                maxArea += max_right - height[right]
                right -= 1
        return maxArea
